### DISCUSSION TITLE
Is there any way to way to concatenate 2 attributes into 1 without using scripts?

### DISCUSSION THREAD

**Discussion by DiegoUrdi** [keycloak/keycloak#22620] [Q&A] [answered]
I'm trying to concatenate two attributes (domain + '.' + id) to send them through the access token, but I can't do it. 

I've tried using the keycloak (v21.0.2) interface, but I don't see any option. 

I've also tried creating next script, but it's not working: 

newvariable-mapper-script.jar:
META-INF\keycloak-scripts.json

```
{
    "authenticators": [],
    "policies": [],
    "mappers": [
        {
            "name": "newVariable Mapper",
            "fileName": "newvariable-mapper.js",
            "description": "Creates the newvariable from the email and the LDAP_ID"
        }
    ],
    "saml-mappers": [] 
}

```

newvariable-mapper.js

```
var emailDomain = user.getSingleAttribute("email");
var ldapId = user.getSingleAttribute("LDAP_ID");

if (emailDomain && ldapId) {
    var concatenatedValue = emailDomain + "." + ldapId;
    user.setAttribute("newVariable", concatenatedValue);
    token.setOtherClaims('newVariable', concatenatedValue);
} else {
    token.setOtherClaims('newVariable', 'concatenatedValue');
}
```

Thanks a lot!!

**Comment by christian-2** [comment]
I don't think there is and you will need either a JavaScript provider (as you tried) or a Java provider (this worked for me in a [similar](https://github.com/keycloak/keycloak/discussions/22487) case) to accomplish this task.

How did you deploy your script (the JAR file)?

**Reply by DiegoUrdi** [reply]
Thanks for your response! 

I deployed it on app.cloud-iam.com, that has an interface to upload the jar directly and redeploys the environment.

I was trying also with the Java provider, but at first glance I have seen it too complex for this that that seemed simpler :(

**Comment by christian-2** [comment]
https://github.com/keycloak/keycloak/discussions/22678 is relevant for you.

**Comment by justin-tay** [comment]
I don't really think there is either. As for your script, the `user.getSingleAttribute()` method doesn't exist so that will throw an exception. I tried the `user.getFirstAttribute()` method which I thought should work but that throws an exception for some reason.

https://github.com/keycloak/keycloak/blob/65d9099e866d214d5ef79c4a3717686ff5bcb7fe/server-spi/src/main/java/org/keycloak/models/UserModel.java#L149

In the end I think the following script should work

```javascript
var emailDomain = user.getAttributeStream("email").findFirst();
var ldapId = user.getAttributeStream("LDAP_ID").findFirst();
if(emailDomain.isPresent() && ldapId.isPresent()) {
    var value = emailDomain.get() + "." + ldapId.get();
    token.setOtherClaims("newVariable", value);
}
```

**Comment by DiegoUrdi** [comment, marked as answer]
Thanks to everyone, finally I could test it locally properly and I had some mistakes. I share them in case it could help someone.

First of all I had to install the nashorn-core-15.4.jar to activate the JavaScript engine on the server. Also I had to install some other libraries (asm and asm-utils).

After that I realized that the .jar structure was wrong. I zipped the whole folder instead of the files directly.

Finally, the email could be read directly from the user.getEmail(), and the LDAP_ID had to be read with getFirstAttribute() as @justin-tay perfectly commented. Also I had to cast it to a string:

```
var email = user.getEmail();
var ldapId = String(user.getFirstAttribute("LDAP_ID"));
```

Thank you very much!!
